### PR TITLE
Val: bindable resource type compatibility between pipeline layout and shader

### PR DESCRIPTION
**Pull request by Jiawei-Shao**
Issue: #3078 

<hr>

**Requirements for PR author:**

- [*] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [*] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [*] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/api/validation/layout_shader_compat.spec.ts [hunk 1]
```diff
@@ -1,14 +1,318 @@
...
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'read-write',
+          },
+        };
+        break;
+      }
+    }
```
Comment (jzm-intel) [inline src/webgpu/api/validation/layout_shader_compat.spec.ts:154]: Or maybe use a dictionary (e.g. `bindGroupLayoutEntryContents`) to map `bindingInPipelineLayout` to the content (`buffer`, `sampler` and others), and we can have a simpler
``` typescript
    const bindGroupLayoutEntryContents = {
      'compareSamp': {
        sampler: {
          type: 'comparison',
        },
      },
      'filtSamp': ...,
      ...
    } as const;

    bindGroupLayoutEntry = {
      binding: 0,
      visibility,
      ...bindGroupLayoutEntryContents[bindingInPipelineLayout],
    };
```

**Review by jzm-intel — APPROVED** [review]
Looks good to me, thanks

File: src/webgpu/api/validation/layout_shader_compat.spec.ts [hunk 2]
```diff
@@ -1,14 +1,318 @@
...
+      .beginSubcases()
+      .combine('pipelineLayoutVisibility', kShaderStageCombinations)
+      .combine('shaderStageWithBinding', kShaderStages)
+      .combine('isBindingStaticallyUsed', [true, false] as const)
+      .unless(
+        p =>
+          // We don't test using non-filtering sampler in shader because it has the same declaration
+          // as filtering sampler.
+          p.bindingInShader === 'nonFiltSamp' ||
+          ((p.bindingInPipelineLayout === 'writeonlyStorageTex' ||
```
Comment (Kangz) [inline src/webgpu/api/validation/layout_shader_compat.spec.ts:214]: nit: Maybe we could have a helper bindingCompatibleWithStage function to make things slightly more readable.

File: src/webgpu/api/validation/layout_shader_compat.spec.ts [hunk 3]
```diff
@@ -1,14 +1,318 @@
...
+            p.bindingInPipelineLayout === 'readwriteStorageTex' ||
+            p.bindingInPipelineLayout === 'storageBuf') &&
+            (p.pipelineLayoutVisibility & GPUConst.ShaderStage.VERTEX) > 0) ||
+          ((p.bindingInShader === 'writeonlyStorageTex' ||
+            p.bindingInShader === 'readwriteStorageTex' ||
+            p.bindingInPipelineLayout === 'storageBuf') &&
```
Comment (Kangz) [inline src/webgpu/api/validation/layout_shader_compat.spec.ts:220]:
```suggestion
            p.bindingInShader === 'storageBuf') &&
```

File: src/webgpu/api/validation/layout_shader_compat.spec.ts [hunk 4]
```diff
@@ -1,14 +1,318 @@
...
+    )}`;
+    const staticallyUseBinding = isBindingStaticallyUsed ? '_ = tmp; ' : '';
+    const isAsync = false;
+    const success =
+      !isBindingStaticallyUsed ||
+      ((pipelineLayoutVisibility & shaderStageWithBinding) > 0 &&
+        (bindingInPipelineLayout === bindingInShader ||
+          (bindingInPipelineLayout === 'nonFiltSamp' && bindingInShader === 'filtSamp') ||
+          (bindingInPipelineLayout === 'readwriteStorageTex' &&
+            bindingInShader === 'writeonlyStorageTex')));
```
Comment (Kangz) [inline src/webgpu/api/validation/layout_shader_compat.spec.ts:245]:
```suggestion
            bindingInShader === 'writeonlyStorageTex')));

```

File: src/webgpu/api/validation/layout_shader_compat.spec.ts [hunk 5]
```diff
@@ -1,14 +1,318 @@
...
+    } = t.params;
+
+    const layout = t.createPipelineLayout(bindingInPipelineLayout, pipelineLayoutVisibility);
+    const bindResourceDeclaration = `@group(0) @binding(0) ${t.GetBindableResourceShaderDeclaration(
+      bindingInShader
+    )}`;
+    const staticallyUseBinding = isBindingStaticallyUsed ? '_ = tmp; ' : '';
+    const isAsync = false;
+    const success =
+      !isBindingStaticallyUsed ||
```
Comment (Kangz) [inline src/webgpu/api/validation/layout_shader_compat.spec.ts:240]: nit: Could we break down this massive expression in subexpressions with names that explain what's going on? Or something like:

```
var success = true;

if (isBindingStaticallyUsed) {
  success = bindingInPipelineLayout === bindingInShader;

  // Filtering and non-filtering both have the same shader declaration.
  success |= bindingInPipelineLayout === 'nonFiltSamp' && bindingInShader === 'filtSamp';

  // Promoting storage textures that are read-write in the layout can be readonly in the shader.
 success |= bindingInPipelineLayout === 'readwriteStorageTex' && bindingInShader === 'writeonlyStorageTex';
 
  // The shader using the resource must be included in the visibility in the layout.
  success &= pipelineLayoutVisibility & shaderStageWithBinding;
}
```

**Review by Kangz — APPROVED** [review]

File: src/webgpu/api/validation/layout_shader_compat.spec.ts [hunk 6]
```diff
@@ -1,14 +1,318 @@
...
+  kShaderStages,
+  ValidBindableResource,
+} from '../../capability_info.js';
+import { GPUConst } from '../../constants.js';
 
 import { ValidationTest } from './validation_test.js';
 
-export const g = makeTestGroup(ValidationTest);
+type BindableResourceType = ValidBindableResource | 'readonlyStorageBuf';
+const kBindableResources = [
```
[the remark this answers is not in the document]
Comment (Jiawei-Shao) [reply src/webgpu/api/validation/layout_shader_compat.spec.ts:19]: Ack

**Review by Kangz — APPROVED** [review]

**Pull request merged**